### PR TITLE
trainer.test() accepts args and kwargs for load_from_checkpoiint()

### LINKED ISSUE #2275: `Add ckpt_path option to LightningModule.test()` introduced robustness issues
## 🐛 Bug

`Trainer.test()`, passing no explicit model, now looks for the best model checkpoint. However, this can fail for models that have `__init__` methods that take more than just `hparams`, which is common.

If I have a `pl` model with init (kw)args `(self, hparams: Namespace, fancy_thing: FancyThing)` one runs into problems outlined below.

With the new codepath, `Trainer.test()` calls down via this path 
```
        # if model is not given (None), ckpt_path is given,
        # load the given checkpoint for testing
        if model is None and ckpt_path is not None:
            # ckpt_path is 'best' so load the best model
            if ckpt_path == 'best':
                ckpt_path = self.checkpoint_callback.best_model_path
            model = self.get_model().load_from_checkpoint(ckpt_path)
```
Which soon calls down to `pytorch_lightning/core/saving.py` @ `_load_model_state`.

This correctly collects `init_args_name` (here `hparams, fancy_thing`) but the subsequent logic does not check whether all the requisite params have been found
```
            if args_name == 'kwargs':
                cls_kwargs = {k: v for k, v in model_args.items() if k in init_args_name}
                kwargs.update(**cls_kwargs)
            elif args_name:
                if args_name in init_args_name:
                    kwargs.update({args_name: model_args})
            else:
                args = (model_args, ) + args

        # load the state_dict on the model automatically
        model = cls(*args, **kwargs)
```
This means the final line of that snippet fails because `fancy_thing` is not passed to `cls`.

Even if it were checked, though, there is not a mechanism here to recover `fancy_thing`. Lightning might need to e.g. pickle the model init args to rehydrate them later, or provide a user hook specifying how the rehydration here should occur. That said, there may be a simple work around I do not see immediately.

### To Reproduce
Call `Trainer.test()` passing no model, where the implicit model has init args that are not just `hparams`.

#### Code sample
I've tried to document very clearly above the code flow that triggers this, but please @ me for any more context.

### Expected behavior

Rehydration should occur properly, or, as a stopgap, an error should be raised explaining that model must be explicitly be passed to `Trainer.test` in this case.

### Additional context

Thanks for the great library; really appreciate the team's work!

### PR DESCRIPTION
**Pull request by yukw777**
## What does this PR do?

Fixes #2275. One thing to note is that now you have to make sure to pass in the model as a kwarg (`trainer.test(model=model)`).

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by yukw777** [comment]
There's no "unreleased" section in the changelog yet so i decided not to touch it yet.

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1101,15 +1101,19 @@ def run_pretrain_routine(self, model: LightningModule):
 
     def test(
             self,
+            *args,
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1104]: this is quite a breaking API change, I would rather have only `**kwargs` and rename it as `loading_kwargs`
cc: @williamFalcon @PyTorchLightning/core-contributors

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:1104]: awesome pr! let's think about this a bit more carefully? mind giving us a few days to recover from the release? haha @yukw777

**Comment by awaelchli** [comment]
Looks really interesting! One thing that comes to mind are these cases:
```python 
trainer.test(1, 2, 3, model=mymodel, kw1='a', kw2='b', kw3='c')
trainer.test(1, 2, 3, model=mymodel, ckpt_path="/path...", kw1='a', kw2='b', kw3='c')
```
Does it load the best checkpoint and uses the new args, or will it ignore the args and use the mymodel reference instead?
In any case it should get documented and perhaps even warn the user?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (yukw777) [reply pytorch_lightning/trainer/trainer.py:1104]: Haha no worries guys, 0.8 was a huge release!! Sleep on this and let me know what you think.

**Comment by yukw777** [comment]
@awaelchli the way i've implemented is that if `model` is passed in, we always use that to test (this made most sense as it's "intentionally" passed in so it should override anything). So in your examples, `mymodel` would be used to test and the best checkpoint or the checkpoint that's passed in is ignored.

**Comment by yukw777** [comment]
@williamFalcon @Borda @awaelchli i see that you guys have been working on fixing `test()` lately, and this approach might be obsolete now? let me know how you guys want to proceed with this!

**Comment by awaelchli** [comment]
I think it can still be useful to override hyperparameters for testing.

**Comment by yukw777** [comment]
I believe this has been fixed by f35337adba0557dddc05bd6b10b896ba68a8e4cf, as we now just take the current model and swap out the weights for testing. No need to reinitialize the model from scratch.

**Pull request closed without merging**